compiler/rustc_hir_typeck/src/pat.rs

```diff
@@ -2309,7 +2340,11 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 }
                 _ => (),
             }
-            if is_slice_or_array_or_vector.0 {
+
+            let enclosure_is_struct =
+                parent_kind.map_or(false, |enclosure| matches!(enclosure, PatKind::Struct(..)));
```

Comment:
Can you try `let enclosure_is_struct = parent_kind.is_some()` instead? I feel like any non-top-level pattern would cause issues. Can you think or a counter-example?

Reply:
I agree with you and fixed.

Reply:
Nice, thank you! I think this is good, I only have a few more comments